File: лабораторная работа 2/SmartDevice.cs
using System;

namespace SmartHomeApp.Models
{
    public abstract class SmartDevice
    {
        private int _id;
        private string _deviceName;
        private bool _isPowered;
        private float _powerConsumption;

        protected SmartDevice(string deviceName, float powerConsumption)
        {
            DeviceName = deviceName;
            PowerConsumption = powerConsumption;
            _isPowered = false;
        }

        protected SmartDevice(int id, string deviceName, float powerConsumption, bool isPowered)
        {
            _id = id;
            DeviceName = deviceName;
            PowerConsumption = powerConsumption;
            _isPowered = isPowered;
        }

        public int Id
        {
            get { return _id; }
            set { _id = value; }
        }

        public string DeviceName
        {
            get { return _deviceName; }
            set 
            { 
                if (string.IsNullOrWhiteSpace(value))
                    _deviceName = "Неизвестное устройство";
                else
                    _deviceName = value;
            }
        }

        public bool IsPowered
        {
            get { return _isPowered; }
            set { _isPowered = value; }
        }

        public float PowerConsumption
        {
            get { return _powerConsumption; }
            set 
            { 
                if (value < 0)
                    _powerConsumption = 0;
                else if (value > 1000)
                    _powerConsumption = 1000;
                else
                    _powerConsumption = value;
            }
        }

        public abstract void TurnOn();
        public abstract void TurnOff();
        public abstract string DisplayInfo();
    }
}

File: лабораторная работа 3/src/main/java/com/vyatsu/smarthome/entity/SmartDevice.java
package com.vyatsu.smarthome.entity;

import javax.persistence.*;

@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "device_type")
public abstract class SmartDevice {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    private String name;
    
    @Column(name = "power_consumption")
    private Double powerConsumption;
    
    @Column(name = "is_powered")
    private Boolean isPowered = false;
    
    // Связь с комнатой (многие к одному)
    @ManyToOne
    @JoinColumn(name = "room_id")
    private Room room;
    
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public Double getPowerConsumption() { return powerConsumption; }
    public void setPowerConsumption(Double powerConsumption) { this.powerConsumption = powerConsumption; }
    
    public Boolean getIsPowered() { return isPowered; }
    public void setIsPowered(Boolean isPowered) { this.isPowered = isPowered; }
    
    public Room getRoom() { return room; }
    public void setRoom(Room room) { this.room = room; }
    
    public abstract void turnOn();
    public abstract void turnOff();
    public abstract String getInfo();
}